How To - App Service to Azure OpenAI Auth (#40568)
dotnet/docs: docs/ai/how-to/app-service-aoai-auth.md

* [Create and deploy an Azure OpenAI Service resource](/azure/ai-services/openai/how-to/create-resource)
* [Create and deploy a .NET application to App Service](/azure/app-service/quickstart-dotnetcore)

## Add a managed identity to App Service

Your application can be granted two types of identities:

* A **system-assigned identity** is tied to your application and is deleted if your app is deleted. An app can only have one system-assigned identity.
* A **user-assigned identity** is a standalone Azure resource that can be assigned to your app. An app can have multiple user-assigned identities.

### Add a system-assigned identity

:::zone target="docs" pivot="azure-portal"

1. Navigate to your app's page on the [Azure Portal](https://aka.ms/azureportal), scroll down to the **Settings** group.
1. Select **Identity**.
1. Within the **System assigned** tab, toggle *Status* to **On**. Click **Save**.

:::zone-end

:::zone target="docs" pivot="azure-cli"

Run the `az webapp identity assign` command to create a system-assigned identity:

```azurecli
az webapp identity assign --name <app-name> --resource-group <group-name>
```

:::zone-end

### Add a user-assigned identity

> To create an app with a user-assigned identity, first create the identity and then add its resource identifier to your app config.

Author: No, it's not a config file it's the app config on Azure

> To create an app with a user-assigned identity, first create the identity and then add its resource identifier to your app config.

Author: We aren't creating the app using a user-assigned identity. We're creating an app that has a user-assigned identity attached to it.

> To create an app with a user-assigned identity, first create the identity and then add its resource identifier to your app config.

Author: Rephrased to:
> To add a user-assigned identity to your app, create the identity, and then add its resource identifier to your app config.

:::zone target="docs" pivot="azure-portal"

1. Create a user-assigned managed identity resource according to these [instructions](/azure/active-directory/managed-identities-azure-resources/how-to-manage-ua-identity-portal#create-a-user-assigned-managed-identity).
1. In the left navigation for your app's page, scroll down to the **Settings** group.
1. Select **Identity**.
1. Select **User assigned > Add**.
1. Search for the identity you created earlier, select it, and then select **Add**.

    > [!CAUTION]

>     > Once you select **Add**, the app restarts.

Author: The impact of the app restarting would depend on the app itself, I'll change this to important because it's up to the dev to decide if that's an issue or not.

:::zone-end

:::zone target="docs" pivot="azure-cli"

1. Create a user-assigned identity:

    ```azurecli
    az identity create --resource-group <group-name> --name <identity-name>
    ```

1. Assign the identity to your app:

    ```azurecli
    az webapp identity assign --resource-group <group-name> --name <app-name> --identities <identity-id>
    ```

:::zone-end

## Add an Azure OpenAI user role to your managed identity

:::zone target="docs" pivot="azure-portal"

1. In the [Azure Portal](https://aka.ms/azureportal), navigate to the scope you want to grant **Azure OpenAI** access to.

>     * The scope can be a **Management group**, **Subscription**, or **Resource group**, or a specific **Azure OpenAI** resource.

Reviewer: ```suggestion
The scope can be a **Management group**, **Subscription**, **Resource group**, or a specific **Azure OpenAI** resource.
```
Did you intend for this to be a bullet item? There's no point to having a sublist item here. If anythng, make this either a note or a second sentence in this list item.

> 1. Select **Access control (IAM)** on the left navigation pane.

Reviewer: ```suggestion
1. In the left navigation pane, select **Access control (IAM)**.
```

1. On the **Members** tab, select the managed identity.
1. On the **Review + assign** tab, select **Review + assign** to assign the role.

:::zone-end

:::zone target="docs" pivot="azure-cli"

**Resource scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "Cognitive Services OpenAI User" \
--scope "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/{providerName}/{resourceType}/{resourceSubType}/{resourceName}"
```

**Resource group scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "Cognitive Services OpenAI User" \
--scope "/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}"
```

**Subscription scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "Cognitive Services OpenAI User" \
--scope "/subscriptions/{subscriptionId}"
```

**Management group scope**

```azurecli
az role assignment create --assignee "{managedIdentityObjectID}" \
--role "Cognitive Services OpenAI User" \
--scope "/providers/Microsoft.Management/managementGroups/{managementGroupName}"
```

:::zone-end

> ## Implement token-based authentication with the Semantic Kernel SDK

Reviewer: ```suggestion
## Implement token-based authentication by using Semantic Kernel SDK
```

> 1. Initialize a `DefaultAzureCredential` object to pick up your app's managed identity:

Reviewer: ```suggestion
1. Initialize a `DefaultAzureCredential` object to assume your app's managed identity:
```
Is this what we mean? Or do we mean "replicate" or something else?

    :::code language="csharp" source="./snippets/semantic-kernel/IdentityExamples.cs" id="tokenCredential":::

> 1. Build a `Kernel` object including the Azure OpenAI Chat Completion Service, and use the previously created credentials:

Reviewer: ```suggestion
1. Build a `Kernel` object that includes the Azure OpenAI Chat Completion Service, and use the previously created credentials:
```
